File: Recursion/N_queen.cpp
#include <bits/stdc++.h>
using namespace std;

// Aspect   	      Complexity   
// Time	              𝑂(𝑁!⋅𝑁)    
// Space (board)  	  𝑂(𝑁2)
// Space (answers)	  𝑂(𝑁!⋅𝑁2)

bool issafe(int row, int col, vector<string>& board, int n) {
    int duprow = row;
    int dupcol = col;

    // Upper-left diagonal
    while (row >= 0 && col >= 0) {
        if (board[row][col] == 'Q') return false;
        row--;
        col--;
    }

    // Left side
    row = duprow;
    col = dupcol;
    while (col >= 0) {
        if (board[row][col] == 'Q') return false;
        col--;
    }

    // Lower-left diagonal
    row = duprow;
    col = dupcol;
    while (row < n && col >= 0) {
        if (board[row][col] == 'Q') return false;
        row++;
        col--;
    }

    return true;
}



void solve(int col, vector<string>& board, vector<vector<string>>& ans, int n) {
    if (col == n) {
        ans.push_back(board);
        return;
    }
    for (int row = 0; row < n; row++) {
        if (issafe(row, col, board, n)) {
            board[row][col] = 'Q';
            solve(col + 1, board, ans, n);
            board[row][col] = '.';
        }
    }
}


// | Metric                             | Complexity                                            |
// | ---------------------------------- | ----------------------------------------------------- |
// | Time                               | `O(N!)`                                               |
// | Space (Without result storage)     | `O(N^2)` (board) + `O(N)` (arrays + stack) = `O(N^2)` |
// | Space (With result storage)        | `O(N! * N^2)` total                                   |

void solve_optimized(int col, vector<string>& board, vector<vector<string>>& ans,vector<int>&leftrow,
    vector<int>&upperdigonal,vector<int>&lowerdiagonal,int n){
        if(col==n){
            ans.push_back(board);
            return;
        }

        for(int row=0;row<n;row++){
            if(leftrow[row]==0 && lowerdiagonal[row+col]==0 && upperdigonal[n-1+col+row]==0){
                board[row][col]='Q';
                leftrow[row]=1;
                lowerdiagonal[row+col]=1;
                upperdigonal[n-1+col-row]=1;
                solve_optimized(col+1,board,ans,leftrow,upperdigonal,lowerdiagonal,n);
                board[row][col]='.';
                leftrow[row]=0;
                lowerdiagonal[row+col]=0;
                upperdigonal[n-1+col-row]=0;
            }
        }
}

int main() {
    int n = 4;
    vector<vector<string>> ans;
    vector<string> board(n, string(n, '.'));
    vector<int>leftrow(n,0),upperdiagonal(2*n-1,0),lowerdiagonal(2*n-1,0);
    solve(0, board, ans, n);
    // Print solutions
    for (auto &sol : ans) {
        for (auto &row : sol)
            cout << row << "\n";
        cout << "\n";
    }
ans.clear(); // <- clear previous solutions

solve_optimized(0, board, ans, leftrow, upperdiagonal, lowerdiagonal, n);
cout << "Optimized Solution:\n";
for (auto &sol : ans) {
    for (auto &row : sol)
        cout << row << "\n";
    cout << "\n";
}
}

File: java/Recursion/NQueen.java
package Recursion;

import java.util.*;

public class NQueen {

    /*
    ====================================================
    TIME & SPACE COMPLEXITY (Overall)
    ----------------------------------------------------
    Time Complexity:
        O(N! * N)
        - N! ways to place queens column-wise
        - Each placement may take O(N) safety checks

    Space Complexity:
        Board            -> O(N^2)
        Recursion Stack  -> O(N)
        Answers Storage  -> O(N! * N^2)
    ====================================================
    */

    /*
    ====================================================
    INTUITION (Brute Force):
    ----------------------------------------------------
    - Place one queen per column.
    - For each cell, check if it is safe:
        1) Upper-left diagonal
        2) Left side (same row)
        3) Lower-left diagonal
    - If safe, place queen and move to next column.
    ====================================================
    */

    static boolean isSafe(int row, int col, char[][] board, int n) {

        int dupRow = row;
        int dupCol = col;

        // Upper-left diagonal
        while (row >= 0 && col >= 0) {
            if (board[row][col] == 'Q') return false;
            row--;
            col--;
        }

        // Left side
        row = dupRow;
        col = dupCol;
        while (col >= 0) {
            if (board[row][col] == 'Q') return false;
            col--;
        }

        // Lower-left diagonal
        row = dupRow;
        col = dupCol;
        while (row < n && col >= 0) {
            if (board[row][col] == 'Q') return false;
            row++;
            col--;
        }

        return true;
    }

    static void solve(int col, char[][] board,
                      List<List<String>> ans, int n) {

        // Base case: all queens placed
        if (col == n) {
            ans.add(construct(board));
            return;
        }

        for (int row = 0; row < n; row++) {
            if (isSafe(row, col, board, n)) {
                board[row][col] = 'Q';
                solve(col + 1, board, ans, n);
                board[row][col] = '.'; // backtrack
            }
        }
    }

    /*
    ====================================================
    OPTIMIZED SOLUTION (Using Hashing)
    ----------------------------------------------------
    - Use 3 arrays to track attacks:
        leftRow[row]
        lowerDiagonal[row + col]
        upperDiagonal[n - 1 + col - row]
    - Safety check becomes O(1)
    ====================================================
    */

    static void solveOptimized(int col, char[][] board,
                               List<List<String>> ans,
                               int[] leftRow,
                               int[] upperDiagonal,
                               int[] lowerDiagonal,
                               int n) {

        if (col == n) {
            ans.add(construct(board));
            return;
        }

        for (int row = 0; row < n; row++) {
            if (leftRow[row] == 0 &&
                lowerDiagonal[row + col] == 0 &&
                upperDiagonal[n - 1 + col - row] == 0) {

                board[row][col] = 'Q';
                leftRow[row] = 1;
                lowerDiagonal[row + col] = 1;
                upperDiagonal[n - 1 + col - row] = 1;

                solveOptimized(col + 1, board, ans,
                               leftRow, upperDiagonal, lowerDiagonal, n);

                // BACKTRACK
                board[row][col] = '.';
                leftRow[row] = 0;
                lowerDiagonal[row + col] = 0;
                upperDiagonal[n - 1 + col - row] = 0;
            }
        }
    }

    // Convert board to list of strings
    static List<String> construct(char[][] board) {
        List<String> res = new ArrayList<>();
        for (char[] row : board) {
            res.add(new String(row));
        }
        return res;
    }

    public static void main(String[] args) {

        int n = 4;
        char[][] board = new char[n][n];

        for (int i = 0; i < n; i++) {
            Arrays.fill(board[i], '.');
        }

        List<List<String>> ans = new ArrayList<>();

        // -------- Brute Force Solution --------
        solve(0, board, ans, n);

        System.out.println("Brute Force Solutions:");
        for (List<String> sol : ans) {
            for (String row : sol)
                System.out.println(row);
            System.out.println();
        }

        // Clear previous results
        ans.clear();

        // -------- Optimized Solution --------
        int[] leftRow = new int[n];
        int[] upperDiagonal = new int[2 * n - 1];
        int[] lowerDiagonal = new int[2 * n - 1];

        solveOptimized(0, board, ans,
                       leftRow, upperDiagonal, lowerDiagonal, n);

        System.out.println("Optimized Solutions:");
        for (List<String> sol : ans) {
            for (String row : sol)
                System.out.println(row);
            System.out.println();
        }
    }
}
